File: go-backend-interview-prep/10_design_patterns_in_go/03_strategy/main.py
"""Strategy pattern -- Python equivalent of the Go example.

Strategies are duck-typed objects with a route() method.
"""


class PriorityRouting:
    """Always picks the first (highest priority) channel."""
    def route(self, message: str, channels: list[str]) -> str:
        return channels[0] if channels else ""


class RoundRobinRouting:
    """Cycles through channels."""
    def __init__(self):
        self.counter = 0

    def route(self, message: str, channels: list[str]) -> str:
        if not channels:
            return ""
        ch = channels[self.counter % len(channels)]
        self.counter += 1
        return ch


class ContentBasedRouting:
    """Picks channel based on message length."""
    def route(self, message: str, channels: list[str]) -> str:
        return "email" if len(message) > 50 else "sms"


class NotificationService:
    def __init__(self, strategy, channels: list[str]):
        self.strategy = strategy
        self.channels = channels

    def send(self, message: str) -> None:
        channel = self.strategy.route(message, self.channels)
        print(f"  Sending via {channel:<8}: {message}")

    def set_strategy(self, strategy) -> None:
        """Swap strategy at runtime."""
        self.strategy = strategy


def main() -> None:
    print("=== Strategy Pattern ===")
    print()

    channels = ["email", "sms", "slack"]

    # Priority strategy
    print("--- Priority routing (always first channel) ---")
    svc = NotificationService(PriorityRouting(), channels)
    svc.send("Order confirmed")
    svc.send("Payment received")
    print()

    # Swap to round-robin
    print("--- Round-robin routing ---")
    svc.set_strategy(RoundRobinRouting())
    svc.send("Order confirmed")
    svc.send("Payment received")
    svc.send("Shipped!")
    svc.send("Delivered!")
    print()

    # Swap to content-based
    print("--- Content-based routing ---")
    svc.set_strategy(ContentBasedRouting())
    svc.send("short msg")
    svc.send("This is a much longer message that should be routed to email because it exceeds fifty characters")
    print()

    print("Key: same service, different routing logic. Swapped at runtime.")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/10_design_patterns_in_go/03_strategy/main.go
package main

import "fmt"

// --- Strategy interface ---

type RoutingStrategy interface {
	Route(message string, channels []string) string
}

// --- Concrete strategies ---

// PriorityRouting always picks the first channel.
type PriorityRouting struct{}

func (p PriorityRouting) Route(message string, channels []string) string {
	if len(channels) == 0 {
		return ""
	}
	return channels[0] // highest priority
}

// RoundRobinRouting cycles through channels.
type RoundRobinRouting struct {
	counter int
}

func (r *RoundRobinRouting) Route(message string, channels []string) string {
	if len(channels) == 0 {
		return ""
	}
	ch := channels[r.counter%len(channels)]
	r.counter++
	return ch
}

// ContentBasedRouting picks channel based on message content.
type ContentBasedRouting struct{}

func (c ContentBasedRouting) Route(message string, channels []string) string {
	if len(message) > 50 {
		return "email" // long messages go to email
	}
	return "sms" // short messages go to sms
}

// --- Service using strategy ---

type NotificationService struct {
	strategy RoutingStrategy
	channels []string
}

func NewNotificationService(strategy RoutingStrategy, channels []string) *NotificationService {
	return &NotificationService{strategy: strategy, channels: channels}
}

func (n *NotificationService) Send(message string) {
	channel := n.strategy.Route(message, n.channels)
	fmt.Printf("  Sending via %-8s: %s\n", channel, message)
}

// SetStrategy allows swapping at runtime.
func (n *NotificationService) SetStrategy(s RoutingStrategy) {
	n.strategy = s
}

func main() {
	fmt.Println("=== Strategy Pattern ===")
	fmt.Println()

	channels := []string{"email", "sms", "slack"}

	// Priority strategy
	fmt.Println("--- Priority routing (always first channel) ---")
	svc := NewNotificationService(PriorityRouting{}, channels)
	svc.Send("Order confirmed")
	svc.Send("Payment received")
	fmt.Println()

	// Swap to round-robin at runtime
	fmt.Println("--- Round-robin routing ---")
	svc.SetStrategy(&RoundRobinRouting{})
	svc.Send("Order confirmed")
	svc.Send("Payment received")
	svc.Send("Shipped!")
	svc.Send("Delivered!")
	fmt.Println()

	// Swap to content-based
	fmt.Println("--- Content-based routing ---")
	svc.SetStrategy(ContentBasedRouting{})
	svc.Send("short msg")
	svc.Send("This is a much longer message that should be routed to email because it exceeds fifty characters")
	fmt.Println()

	fmt.Println("Key: same service, different routing logic. Swapped at runtime.")
}
